core/src/main/java/io/grpc/MessageEncoding.java

```diff
@@ -108,36 +118,93 @@ public InputStream decompress(InputStream is) throws IOException {
 
   /**
    * Registers a decompressor for both decompression and message encoding negotiation.
+   *
+   * @param d The decompressor to register
+   * @param advertised If true, the message encoding will be listed in the Accept-Encoding header.
    * @throws IllegalArgumentException if another compressor by the same name is already registered.
    */
-  public static final void registerDecompressor(Decompressor d) {
-    Decompressor previousDecompressor = decompressors.putIfAbsent(d.getMessageEncoding(), d);
-    checkArgument(previousDecompressor == null,
-        "A decompressor was already registered: %s", previousDecompressor);
+  @ExperimentalApi
+  public static final void registerDecompressor(Decompressor d, boolean advertised) {
+    DecompressorInfo previousInfo =
+        decompressors.putIfAbsent(d.getMessageEncoding(), new DecompressorInfo(d, advertised));
+    if (previousInfo != null) {
+      throw new IllegalArgumentException(
+          "A decompressor was already registered: " + previousInfo.decompressor);
+    }
   }
 
   /**
    * Provides a list of all message encodings that have decompressors available.
    */
-  public static Collection<String> getKnownMessageEncodings() {
+  @ExperimentalApi
+  public static Set<String> getKnownMessageEncodings() {
```

Comment:
I think the name of the class `MessageEncoding` is getting to be confusing.  We have a few different concepts here that I feel are getting conflated:

1. The Codec API
2. Default implementations of Codec
3. Codec registry

For 1, I think that this might benefit from extracting the `Codec` API out to a top-level interface (i.e separate it from the registry).

For 2, as discussed above we could move the `NONE` codec to either the `Codec` interface or to a `Codecs` utility class.

For 3, I think we should create a class called `CodecRegistry` ... perhaps rename `MessageEncoding`.

WDYT?